src/hotspot/share/opto/phaseX.cpp

```diff
@@ -2285,6 +2289,44 @@ void PhasePeephole::print_statistics() {
 }
 #endif
 
+Node* PhaseLowering::apply_ideal(Node* k, bool can_reshape) {
+  Node* lowered = lower_node(k);
+  if (lowered != nullptr) {
+    return lowered;
+  }
+
+  // Run the lowered Ideal method to continue doing transformations on the node, while avoiding existing ideal transforms
+  // that may undo the changes done during lowering.
+
+  return k->LoweredIdeal(this);
+}
+
+Node* PhaseLowering::apply_identity(Node* k) {
+  return k;
+}
+
+Node* PhaseLowering::lower_node(Node* n) {
+  // Apply shared lowering transforms
+
```

Comment:
Per my understanding, this is a backend specific lowering phase, is there any scenario that a platform in-dependent lowering is needed here? As we already have the common GVN phase for common node idealize, is there any difference for such shared transformations? Thanks!

Reply:
Lowering is not idealisation so I think having backend independent lowering is fine in case we need it.

Reply:
This is the only place to do transformations with GVN but without Ideal so I think the distinction is clear enough to avoid potential confusion.